### DISCUSSION TITLE
How to use EFQRCode in SwiftUI

### DISCUSSION THREAD

**Discussion by ApolloZhu** [EFPrefix/EFQRCode#129] [Q&A] [answered]
Currently, there are no examples in, nor native support for SwiftUI

**Comment by ApolloZhu** [comment, marked as answer]
EFQRCode generates a `CGImage` on success, so ideally we should be able to make a SwiftUI `Image` from the result like this:

```swift
struct ContentView: View {
    var body: some View {
        if let qrCode = EFQRCode.generate(for: "Hello World") {
            Image(cgImage: qrCode)
                .resizable()
                .aspectRatio(1, contentMode: .fit)
        } else {
            Text("Generation Failed")
        }
    }
}
```

However, since SwiftUI doesn't support doing that, we need to add the following initializer extension ourselves:

```swift
extension Image {
    init(cgImage: CGImage) {
        #if os(macOS)
        self.init(nsImage: NSImage(cgImage: cgImage,
                                   size: NSSize(width: cgImage.width,
                                                height: cgImage.height)))
        #else
        self.init(uiImage: UIImage(cgImage: cgImage))
        #endif
    }
}
```

Then, you can display the generated QRCode with SwiftUI (on any platform supported by SwiftUI and EFQRCode, such as watchOS):

<img width="1348" alt="the above code snippets in Xcode, alongside with the generated QRCode image shown in a watchOS simulator" src="https://user-images.githubusercontent.com/10842684/128489420-e065e94d-3898-4c4e-8bf6-7a02894ee22d.png">

---

> Disclaimer: It might (or may not) be a better idea to use `@State` and `onAppear` to cache the generated QRCode?

**Reply by AhmadKsb** [reply]
Thank you for this!
